File: packages/sandboxd/src/sandboxd_state/runtime_coordination.rs
use std::sync::Arc;
use std::sync::atomic::{AtomicBool, Ordering};
use std::thread::{self, JoinHandle};

use crate::codex_proxy::CodexProxyControlHandle;
use crate::process::{CodexAppServerControlHandle, OpenCodeServerControlHandle};
use crate::supervision::{SandboxdSupervisorHandle, SupervisedComponent};

const RUNTIME_COORDINATION_POLL_INTERVAL: std::time::Duration =
    std::time::Duration::from_millis(250);
const CODEX_PROXY_RECOVERY_TIMEOUT: std::time::Duration = std::time::Duration::from_secs(5);

#[derive(Clone, Debug, Default)]
pub(super) struct RuntimeCoordinationHandles {
    pub(super) codex_app_server_control_handle: Option<CodexAppServerControlHandle>,
    pub(super) codex_proxy_control_handle: Option<CodexProxyControlHandle>,
    pub(super) opencode_server_control_handle: Option<OpenCodeServerControlHandle>,
}

impl RuntimeCoordinationHandles {
    pub(super) fn has_runtime_process_control(&self) -> bool {
        self.opencode_server_control_handle.is_some()
            || (self.codex_app_server_control_handle.is_some()
                && self.codex_proxy_control_handle.is_some())
    }
}

pub(super) fn spawn_runtime_coordination_thread(
    handles: RuntimeCoordinationHandles,
    supervisor_handle: SandboxdSupervisorHandle,
    shutdown_requested: Arc<AtomicBool>,
) -> JoinHandle<()> {
    thread::spawn(move || {
        run_runtime_coordination_loop(handles, supervisor_handle, shutdown_requested);
    })
}

fn run_runtime_coordination_loop(
    handles: RuntimeCoordinationHandles,
    supervisor_handle: SandboxdSupervisorHandle,
    shutdown_requested: Arc<AtomicBool>,
) {
    while !shutdown_requested.load(Ordering::Relaxed) {
        if let (Some(codex_app_server_control_handle), Some(codex_proxy_control_handle)) = (
            &handles.codex_app_server_control_handle,
            &handles.codex_proxy_control_handle,
        ) {
            coordinate_codex_runtime(
                codex_app_server_control_handle,
                codex_proxy_control_handle,
                &supervisor_handle,
                shutdown_requested.as_ref(),
            );
        }
        if let Some(opencode_server_control_handle) = &handles.opencode_server_control_handle {
            coordinate_opencode_runtime(opencode_server_control_handle, &supervisor_handle);
        }
        thread::sleep(RUNTIME_COORDINATION_POLL_INTERVAL);
    }
}

fn coordinate_codex_runtime(
    codex_app_server_control_handle: &CodexAppServerControlHandle,
    codex_proxy_control_handle: &CodexProxyControlHandle,
    supervisor_handle: &SandboxdSupervisorHandle,
    shutdown_requested: &AtomicBool,
) {
    let Some(codex_app_server_snapshot) =
        supervisor_handle.component_snapshot(SupervisedComponent::CodexAppServer)
    else {
        return;
    };
    if codex_app_server_snapshot.state != crate::supervision::ComponentHealthState::Restarting {
        return;
    }

    let restart_reason = component_restart_reason(&codex_app_server_snapshot);
    supervisor_handle.emit_component_restart_scheduled(
        SupervisedComponent::CodexAppServer,
        restart_reason,
        0,
        &[],
    );

    if codex_app_server_control_handle
        .restart(&crate::time::SystemClock, &crate::time::ThreadSleeper)
        .is_ok()
        && !wait_for_codex_proxy_recovery(
            codex_proxy_control_handle,
            CODEX_PROXY_RECOVERY_TIMEOUT,
            shutdown_requested,
        )
    {
        let _ = codex_proxy_control_handle.request_restart();
        let _ = wait_for_codex_proxy_recovery(
            codex_proxy_control_handle,
            CODEX_PROXY_RECOVERY_TIMEOUT,
            shutdown_requested,
        );
    }
}

fn coordinate_opencode_runtime(
    opencode_server_control_handle: &OpenCodeServerControlHandle,
    supervisor_handle: &SandboxdSupervisorHandle,
) {
    let Some(opencode_server_snapshot) =
        supervisor_handle.component_snapshot(SupervisedComponent::OpenCodeServer)
    else {
        return;
    };
    if opencode_server_snapshot.state != crate::supervision::ComponentHealthState::Restarting {
        return;
    }

    let restart_reason = component_restart_reason(&opencode_server_snapshot);
    supervisor_handle.emit_component_restart_scheduled(
        SupervisedComponent::OpenCodeServer,
        restart_reason,
        0,
        &[],
    );

    let _ = opencode_server_control_handle
        .restart(&crate::time::SystemClock, &crate::time::ThreadSleeper);
}

fn component_restart_reason(
    snapshot: &crate::supervision::ComponentHealthSnapshot,
) -> &'static str {
    if snapshot
        .details
        .get("livenessState")
        .is_some_and(|liveness_state| liveness_state == "Exited")
    {
        "coordinated_restart_after_exit"
    } else {
        "coordinated_restart_after_readiness_failure"
    }
}

fn wait_for_codex_proxy_recovery(
    codex_proxy_control_handle: &CodexProxyControlHandle,
    timeout: std::time::Duration,
    shutdown_requested: &AtomicBool,
) -> bool {
    let deadline = std::time::Instant::now() + timeout;
    loop {
        if shutdown_requested.load(Ordering::Relaxed) {
            return false;
        }
        if let Some(snapshot) = codex_proxy_control_handle.snapshot() {
            let raw_connectivity_connected = snapshot
                .details
                .get("rawConnectivityState")
                .is_some_and(|state| state == "Connected");
            let session_manager_connected = snapshot
                .details
                .get("sessionManagerState")
                .is_some_and(|state| state == "Connected");
            if snapshot.state == crate::supervision::ComponentHealthState::Healthy
                && raw_connectivity_connected
                && session_manager_connected
            {
                return true;
            }
        }
        if std::time::Instant::now() >= deadline {
            return false;
        }
        thread::sleep(RUNTIME_COORDINATION_POLL_INTERVAL);
    }
}

File: packages/sandboxd/sandboxdstate/runtime_coordination.go
package sandboxdstate

import (
	"sync"
	"time"

	"github.com/mistle/sandboxd/process"
	"github.com/mistle/sandboxd/supervision"
	"github.com/mistle/sandboxd/timeutil"
)

const (
	RuntimeCoordinationPollInterval = 250 * time.Millisecond
	CodexProxyRecoveryTimeout       = 5 * time.Second
)

type RuntimeCoordinationHandles struct {
	CodexAppServerControlHandle *process.CodexAppServerControlHandle
	CodexProxyControlHandle     *CodexProxyControlHandle
	OpenCodeServerControlHandle *process.OpenCodeServerControlHandle
}

type RuntimeCoordinationHandle struct {
	shutdown chan struct{}
	done     chan struct{}
	once     sync.Once
}

func (handles RuntimeCoordinationHandles) HasRuntimeProcessControl() bool {
	return handles.OpenCodeServerControlHandle != nil ||
		(handles.CodexAppServerControlHandle != nil && handles.CodexProxyControlHandle != nil)
}

func StartRuntimeCoordination(
	handles RuntimeCoordinationHandles,
	supervisorHandle *supervision.SandboxdSupervisorHandle,
	clock timeutil.Clock,
	sleeper timeutil.Sleeper,
) *RuntimeCoordinationHandle {
	handle := &RuntimeCoordinationHandle{
		shutdown: make(chan struct{}),
		done:     make(chan struct{}),
	}
	go func() {
		defer close(handle.done)
		runRuntimeCoordinationLoop(handles, supervisorHandle, clock, sleeper, handle.shutdown)
	}()
	return handle
}

func (handle *RuntimeCoordinationHandle) Close() {
	if handle == nil {
		return
	}
	handle.once.Do(func() {
		close(handle.shutdown)
		<-handle.done
	})
}

func runRuntimeCoordinationLoop(
	handles RuntimeCoordinationHandles,
	supervisorHandle *supervision.SandboxdSupervisorHandle,
	clock timeutil.Clock,
	sleeper timeutil.Sleeper,
	shutdown <-chan struct{},
) {
	for !coordinationShutdownRequested(shutdown) {
		coordinateCodexRuntime(handles.CodexAppServerControlHandle, handles.CodexProxyControlHandle, supervisorHandle, clock, sleeper, shutdown)
		coordinateOpenCodeRuntime(handles.OpenCodeServerControlHandle, supervisorHandle, clock, sleeper)
		sleeper.Sleep(RuntimeCoordinationPollInterval)
	}
}

func coordinateCodexRuntime(
	codexAppServerControlHandle *process.CodexAppServerControlHandle,
	codexProxyControlHandle *CodexProxyControlHandle,
	supervisorHandle *supervision.SandboxdSupervisorHandle,
	clock timeutil.Clock,
	sleeper timeutil.Sleeper,
	shutdown <-chan struct{},
) {
	if codexAppServerControlHandle == nil || codexProxyControlHandle == nil {
		return
	}
	snapshot := supervisorHandle.ComponentSnapshot(supervision.ComponentCodexAppServer)
	if snapshot == nil || snapshot.State != supervision.ComponentRestarting {
		return
	}

	supervisorHandle.EmitComponentRestartScheduled(
		supervision.ComponentCodexAppServer,
		componentRestartReason(snapshot),
		0,
		nil,
	)
	if err := codexAppServerControlHandle.Restart(clock, sleeper); err != nil {
		return
	}
	if waitForCodexProxyRecovery(codexProxyControlHandle, CodexProxyRecoveryTimeout, shutdown) {
		return
	}
	if err := codexProxyControlHandle.RequestRestart(); err != nil {
		return
	}
	_ = waitForCodexProxyRecovery(codexProxyControlHandle, CodexProxyRecoveryTimeout, shutdown)
}

func coordinateOpenCodeRuntime(
	openCodeServerControlHandle *process.OpenCodeServerControlHandle,
	supervisorHandle *supervision.SandboxdSupervisorHandle,
	clock timeutil.Clock,
	sleeper timeutil.Sleeper,
) {
	if openCodeServerControlHandle == nil {
		return
	}
	snapshot := supervisorHandle.ComponentSnapshot(supervision.ComponentOpenCodeServer)
	if snapshot == nil || snapshot.State != supervision.ComponentRestarting {
		return
	}

	supervisorHandle.EmitComponentRestartScheduled(
		supervision.ComponentOpenCodeServer,
		componentRestartReason(snapshot),
		0,
		nil,
	)
	_ = openCodeServerControlHandle.Restart(clock, sleeper)
}

func componentRestartReason(snapshot *supervision.ComponentHealthSnapshot) string {
	if snapshot.Details["livenessState"] == "Exited" {
		return "coordinated_restart_after_exit"
	}
	return "coordinated_restart_after_readiness_failure"
}

func waitForCodexProxyRecovery(
	codexProxyControlHandle *CodexProxyControlHandle,
	timeout time.Duration,
	shutdown <-chan struct{},
) bool {
	deadline := time.Now().Add(timeout)
	for {
		if coordinationShutdownRequested(shutdown) {
			return false
		}
		snapshot := codexProxyControlHandle.Snapshot()
		if snapshot != nil &&
			snapshot.State == supervision.ComponentHealthy &&
			snapshot.Details["rawConnectivityState"] == string(codexproxyHealthConnected) &&
			snapshot.Details["sessionManagerState"] == string(codexproxyHealthConnected) {
			return true
		}
		if time.Now().After(deadline) {
			return false
		}
		time.Sleep(RuntimeCoordinationPollInterval)
	}
}

func coordinationShutdownRequested(shutdown <-chan struct{}) bool {
	select {
	case <-shutdown:
		return true
	default:
		return false
	}
}

type codexproxyHealthState string

const codexproxyHealthConnected codexproxyHealthState = "Connected"
